### PR TITLE
dtv/atsc: replace structs with streams and tags

### PR DESCRIPTION
**Pull request by mormj**
ATSC implementation is old, and I don't know the entire history of it. But it seems to be the only module that uses structs to pass data between blocks.  This causes the grc blocks to be very special purpose and only have byte vectors as the interface (which many are floats).

The other reasoning behind this is also trying to keep packet data in contiguous memory so that hardware accelerators can be used more easily on larger buffer sizes

### CODE REVIEW

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ */
+class plinfo
+{
+public:
+    plinfo() : _flags(0), _segno(0) {}
+    // plinfo(uint16_t flags_, int16_t segno_) : _flags(flags_), _segno(segno_) {}
+
+    // accessors
+
+    bool field_sync1_p() const { return (_flags & fl_field_sync1) != 0; }
```
Comment (ThomasHabets) [inline gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:34]: Since this class is now part of the public API (and even ABI) it could make sense to clean it up a bit.

E.g.:
* Add `noexcept` where appropriate
* change state machine (?) to `enum class`
* use member initialization
* remove `operator==/!=` (they're default, no?
* Make member variables private?
* Put implementation in a .cc in `lib/`

What do you think?

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc [hunk 2]
```diff
@@ -148,20 +163,39 @@ int atsc_equalizer_impl::general_work(int noutput_items,
...
             filterN(data_mem, data_mem2, ATSC_DATA_SEGMENT_LENGTH);
 
-            memcpy(out[output_produced].data,
+            memcpy(&out[output_produced * ATSC_DATA_SEGMENT_LENGTH],
                    data_mem2,
                    ATSC_DATA_SEGMENT_LENGTH * sizeof(float));
 
-            out[output_produced].pli._flags = d_flags;
-            out[output_produced].pli._segno = d_segno;
+            // out[output_produced].pli._flags = d_flags;
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_equalizer_impl.cc:170]: commented out code?

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc [hunk 3]
```diff
@@ -148,20 +163,39 @@ int atsc_equalizer_impl::general_work(int noutput_items,
...
+            // out[output_produced].pli._segno = d_segno;
+            plinfo pli_out;
+            pli_out._flags = d_flags;
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_equalizer_impl.cc:173]: What about endianness if using multiple hosts, or merely replaying a capture file?

**Review by ThomasHabets — CHANGES_REQUESTED** [review]
It is weird that ATSC sends serialized structs, yes. But maybe it's worth breaking out the data portions of the tag into a just-the-data struct? (endian safe)

File: gr-dtv/lib/atsc/atsc_fs_checker_impl.cc [hunk 4]
```diff
@@ -59,17 +59,20 @@ int atsc_fs_checker_impl::general_work(int noutput_items,
                                        gr_vector_const_void_star& input_items,
                                        gr_vector_void_star& output_items)
 {
-    const atsc_soft_data_segment* in = (const atsc_soft_data_segment*)input_items[0];
-    atsc_soft_data_segment* out = (atsc_soft_data_segment*)output_items[0];
+    const float* in = reinterpret_cast<const float*>(input_items[0]);
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_fs_checker_impl.cc:62]: Is `static_cast<>` not enough?

File: gr-dtv/lib/atsc/atsc_sync_impl.cc [hunk 5]
```diff
@@ -82,15 +82,15 @@ int atsc_sync_impl::general_work(int noutput_items,
                                  gr_vector_const_void_star& input_items,
                                  gr_vector_void_star& output_items)
 {
-    const float* in = (const float*)input_items[0];
-    atsc_soft_data_segment* soft_data_segment_out =
-        (atsc_soft_data_segment*)output_items[0];
+    const float* in = reinterpret_cast<const float*>(input_items[0]);
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_sync_impl.cc:85]: same

File: gr-dtv/lib/atsc/atsc_sync_impl.cc [hunk 6]
```diff
@@ -180,8 +180,13 @@ int atsc_sync_impl::general_work(int noutput_items,
             d_data_mem[d_symbol_index] = interp_sample;
 
             if (d_symbol_index >= (ATSC_DATA_SEGMENT_LENGTH - 1)) {
-                for (int i = 0; i < ATSC_DATA_SEGMENT_LENGTH; i++)
-                    soft_data_segment_out[d_output_produced].data[i] = d_data_mem[i];
+                // for (int i = 0; i < ATSC_DATA_SEGMENT_LENGTH; i++)
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_sync_impl.cc:183]: Should this commented-out code remain?

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+namespace gr {
+namespace dtv {
+
+/*!
+ * \brief pipeline info that flows with data
+ *
+ * Not all modules need all the info
+ */
+
+struct plinfo_data {
```
Comment (ThomasHabets) [inline gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:27]: so this is a new one, but it's not used by `plinfo`? I'm not sure I follow.

**Review by ThomasHabets — CHANGES_REQUESTED** [review]

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h (hunk 7 above)
Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:27]: The idea here was to break out the data portion of the class into a struct.  It is only this portion that gets sent along on the tags now.

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc (hunk 3 above)
Comment (mormj) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc]: I'm not sure I follow - is this in reference to a struct being sent over std::any?

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h (hunk 1 above)
Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:44]: Agreed - I have more work to clean up this class

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc (hunk 3 above)
Comment (ThomasHabets) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc]: No. The binary representation of the tag value is not portable across little/big endian. So saving a stream on a big endian machine it won't be correctly readable on a little endian machine unless they're encoded somehow. Either `htonl()` and friends, or some other serialization.

As I see it now the values are just used in the data stream as whatever the native endianness is.

**Comment by willcode** [comment]
If atsc_plinfo.h is in include/gnuradio/dtv, it (and the functions in it) become part of the API. Is that the intent? This file could be split into public/impl or just moved to lib if not.

**Comment by mormj** [comment]
Yes, the intent is that atsc_plinfo.h public, since a user could break off from the receive chain and insert, say, hardware accelerated receive blocks that are in an OOT that would need access to the struct in  atsc_plinfo.h to decode tags

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h (hunk 1 above)
Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: Ok, I've done the things here except for the `noexcept` - would like to see this applied consistently through the codebase and it is currently mainly used for `noexcept(false)`

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc (hunk 3 above)
Comment (mormj) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc]: Sending a big endian packed U32 with the two values over the tags now.

File: gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc [hunk 8]
```diff
@@ -52,24 +55,37 @@ int atsc_deinterleaver_impl::work(int noutput_items,
...
+    auto tag_pmt = pmt::intern("plinfo");
     for (int i = 0; i < noutput_items; i++) {
-        assert(in[i].pli.regular_seg_p());
-        plinfo::sanity_check(in[i].pli);
+        plinfo pli_in;
+        get_tags_in_window(tags, 0, i, i + 1, tag_pmt);
+        if (tags.size() > 0) {
+            pli_in.from_tag_value(pmt::to_uint64(tags[0].value));
+        } else {
+            throw std::runtime_error("No plinfo on tag");
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc:69]: On "no plinfo tag on sample", you mean?

File: gr-dtv/lib/atsc/atsc_plinfo.cc [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+void plinfo::reset()
+{
+    _flags = 0;
+    _segno = 0;
+}
+
+void plinfo::from_tag_value(uint32_t tag_value)
+{
+    auto native_val = boost::endian::big_to_native(tag_value);
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_plinfo.cc:27]: Why not `htonl()`, to avoid dependency on boost?

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+    // This bit is set when Reed-Solomon decoding detects an error that it
+    // can't correct.  Note that other error detection (e.g. Viterbi) do not
+    // set it, since Reed-Solomon will correct many of those.  This bit is
+    // then copied into the final Transport Stream packet so that MPEG
+    // software can see that the 188-byte data segment has been corrupted.
+    static constexpr int fl_transport_error = 0x0020;
+
+private:
+    uint16_t _flags = 0; // bitmask
```
Comment (ThomasHabets) [inline gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:110]: With it being moved to public headers, is this a good time to make it conform to naming standard with `d_flags`?

**Review by ThomasHabets — APPROVED** [review]

File: gr-dtv/lib/atsc/atsc_plinfo.cc (hunk 9 above)
Comment (mormj) [reply gr-dtv/lib/atsc/atsc_plinfo.cc:27]: We already have the dependency on boost::endian via pmt, which is already cross platform.  htonl is very reasonable but require the inclusion of different headers depending on Linux/Windows/Mac, and hasn't yet been used in the codebase.

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h (hunk 10 above)
Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:110]: good point.

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+     * @param tag_value
+     */
+    void from_tag_value(uint32_t tag_value);
+
+    /**
+     * @brief Return an endian safe value containing the flags and segno
+     *
+     * @return uint32_t
+     */
+    uint32_t get_tag_value();
```
Comment (ThomasHabets) [inline gr-dtv/include/gnuradio/dtv/atsc_plinfo.h:51]: `const`

**Review by ThomasHabets — APPROVED** [review]

**Review by ThomasHabets — APPROVED** [review]

**Review by drmpeg — APPROVED** [review]
Tested live over the air. Looks good! Maybe takes slightly more CPU cycles, but nothing to worry about.

**Comment by mormj** [comment]
Doesn't make a functional difference, but added the `DTV_API` tag to the plinfo class for consistency (cross-platform ramifications?)

File: gr-dtv/include/gnuradio/dtv/atsc_plinfo.h (hunk 1 above)
Comment (mbr0wn) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: There's something with `noexcept` and new C++ standards. Something changed -- @ThomasHabets would know best.

Comment (ThomasHabets) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: `noexcept(false)` is the default. I'm not sure if you're thinking of the now deprecated `throw()` spec. Except when writing templates that can instantiatie to `noexcept` or not depending on template args, there's no need for parentheses.

`throw()` was always a mess, unfortunately. `noexcept` is fine though. `throw()` ended up being only used for empty throw specs.

In my testing `noexcept` reduces code size. I've not benchmarked but it can enable other optimizations beyond reduced instruction cache use.

But the main reason I suggested it is that it's a nice annotation to the reader that this function cannot even in principle throw. I'm pretty sure it's not part of the ABI, but for public APIs better safe than sorry so should only be added when no possible future implementation could throw, in my opinion.

Of course in this case it'll in practice be inlined anyway, and the compiler will not generate exception handling code on entry or exit.

Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: I definitely trust @ThomasHabets on this one.  Pushed a commit that puts `noexcept` _everywhere_ in the plinfo class

Comment (mormj) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: On second thought, I don't think we're ready to move in the direction of putting `noexcept` in lots of places, and probably need a consistent strategy for exceptions and exception handling.  Backing out the last commit to keep this file consistent with the rest of the codebase

Comment (ThomasHabets) [reply gr-dtv/include/gnuradio/dtv/atsc_plinfo.h]: Thinking first and rolling out a consistent use based on clear criteria sounds best, I agree.

**Comment by mbr0wn** [comment]
CentOS failure is unrelated.

**Comment by drmpeg** [comment]
I'm afraid this version came out pretty glitchy. Somethings not right somewhere, but it's going to take a while for me to track down the offending block(s).

**Comment by mormj** [comment]
Darn.  Is it glitchy in terms of decoded video or in some other way?
Nothing other than a squash and the DTV_API attribute changed.

**Comment by drmpeg** [comment]
Yes, glitchy decoded video. Looks like the ATSC Receiver SYNC block is having problems. I'm uploading a test file right now, but I only have 12 Mbps upstream on my cable Internet. I'll let you know when it's ready.

I'm not sure why I didn't catch this. When I tested before, I did see some glitches at one point, and then they went away. Maybe I got fooled with the way I was switching between versions. I had two builds and was just using ```sudo make install`` to switch between the two. Today, I have one build, and I'm switching commits and rebuilding (and installing).

**Comment by mormj** [comment]
Thanks for digging in - Looks like I've got some debugging to do.  If I can't figure out we'll obviously have to back out this commit.

**Comment by drmpeg** [comment]
Doh! I just realized I can't compare block outputs between the two versions without taking into account the new structures. So it may not be the ATSC Receiver SYNC block.

**Comment by drmpeg** [comment]
Yup, ATSC Receiver SYNC block is okay when the header and padding on the old receiver is discarded.

**Comment by drmpeg** [comment]
Okay, I'm pretty sure it's the ATSC Viterbi Decoder block. It fails pretty far into the test bitstream after 690,552 output bytes, which is exactly 3,336 207-byte packets.

Then it recovers after byte 693,035, which is 2,484 bytes or 12 207-byte packets later.

Then it fails again after 1,428,300 bytes or 6,900 207-byte packets (and recovers after 2484 bytes).

1,614,600 or 7,800 packets
3,174,552 or 15,336 packets
4,192,992 or 20,256 packets
4,523,364 or 21,852 packets
5,412,636 or 26,148 packets
8,793,360 or 42,480 packets

The first failing block had all zeroes in it, so it seems that 12 207-byte packets don't get written out after some corner case.

**Comment by drmpeg** [comment]
The test bitstream is ready. It's 5 minutes, so it's pretty ginormous. 28,451,236,144 bytes.

http://www.w6rz.net/atscsator.cfile

New test flow graph:

image

Old test flow graph:

image

```cmp -l atscvitold.bin atscvitnew.bin | more```

**Comment by drmpeg** [comment]
Although the ATSC decoder is deterministic, the bug isn't. You'll see different error offsets than what I posted above and different offsets from run to run.

**Comment by mormj** [comment]
@drmpeg - thank you for setting up the debug so well.  I can't seem to find any arithmetic errors in the inputs to the viterbi decoder or copying the outputs just yet - most of the code is the same apart from the plinfo and copying into the various structs.  

One thing I do notice is that the plinfo coming into the viterbi decoder is off by one sample which could cause issues - not sure whether this would cause non-deterministic issues

Now, I'm wondering though whether it would be better to pass the plinfo as a separate stream of data rather than tags - since it is always one per sample.

**Comment by mormj** [comment]
Nope, it is not the plinfo - eliminated that variable - strictly something happening in the viterbi decoder :/
